### DISCUSSION TITLE
Future of VegaFusion and Altair

### DISCUSSION THREAD

**Discussion by jonmmease** [vega/altair#3054] [Ideas]
Now that Altair 5 is out (thanks again to @mattijn @joelostblom and @binste for all of the work to make this a reality!), I wanted to start a discussion on the future relationship between VegaFusion and Altair.  This is also in response to @joelostblom's question in https://github.com/altair-viz/altair/issues/3053.

Even though VegaFusion is a cool piece of technology, it's not really something an end user of Altair should need to think about. From my perspective, It would be great if the capabilities added by VegaFusion were just seen as capabilities of Altair itself to the casual user.

Today, the VegaFusion Python library consists of two parts. A low-level runtime that deals exclusively with dictionaries of Vega specs and has no dependency on Altair (e.g. `vf.runtime.pre_transform_spec`). And higher level entry points that integrate with Altair using vl-convert (e.g. the mime renderer, save functions, and `transformed_data` function).  

A question for the Altair community is whether we would be interested in making VegaFusion an *optional* dependency and then port this high-level functionality into Altair itself. 

### Example Integrations
Here are a couple of examples of what the more integrated functionality could look like:

The existing renderers could get a `vegafusion` kwarg to control whether serverside transformation is used (and it could default to `True` if VegaFusion is installed).
```python
alt.renderers.enable("mimetype", vegafusion=True)
```

or for the VegaFusion widget renderer
```python
alt.renderers.enable("widget").
```

Similarly, the `Chart.save` method could have a `vegafusion` kwarg to control whether transformations are applied before saving the chart:
```python
Chart.save("chart.html", vegafusion=True)
```

Charts could have a `.transformed_data()` method alongside their existing `.data` property to access the chart's data post transformation:

```python
chart.transformed_data()
```

And the `vegafusion[embed]` package specification could be added to an `extra_requires` group to install automatically with `pip install altair[full]`.

### Always optional
As with vl-convert, it's important that a VegaFusion dependency always be optional since VegaFusion is a compiled library and won't be available everywhere that Altair currently works (e.g. in pyodide).

### Development and Maintenance
If we want to go this direction, I'd be happy to do the bulk of the integration work and to join as an Altair maintainer.

**Comment by joelostblom** [comment]
> From my perspective, It would be great if the capabilities added by VegaFusion were just seen as capabilities of Altair itself to the casual user.

I agree with this, I think it is important that the out of the box experience includes the most crucial conveniences for users and I think smaller sized charts by default is one of these and aligns closely with what the user would likely expect when making a chart (i. e. only the data required to display the chart is included in the spec unless explicitly instructed to include all data)

> The existing renderers could get a vegafusion kwarg to control whether serverside transformation is used (and it could default to True if VegaFusion is installed).

I think this sounds like a good solution to me. I wonder if the kwarg should be called `vegafusion` or something more descriptive like `transform_data` (but it might be hard to find a name that encompasses everything that vegafusion does and that might also evolve over time, so `vegafusion` might be the safest choice).

> A question for the Altair community is whether we would be interested in making VegaFusion an optional dependency and then port this high-level functionality into Altair itself... If we want to go this direction, I'd be happy to do the bulk of the integration work and to join as an Altair maintainer.

On the top of my head, I would be in favor of this. Could you give a rough indication of the amount of code that would be integrated into Altair, is it a few hundred lines or thousands? And thank you for offering to do the maintenance of this part of the code for the library, I personally think that sounds like the ideal solution.


> And the vegafusion[embed] package specification could be added to an extra_requires group to install automatically with pip install altair[full].

I think this would be great, and having a similar conda package would be convenient as well, but I am not sure if conda supports this functionality now or if we would need to create a meta package.

> As with vl-convert, it's important that a VegaFusion dependency always be optional since VegaFusion is a compiled library and won't be available everywhere that Altair currently works (e.g. in pyodide).

Thanks for  noting this, makes sense!

**Reply by jonmmease** [reply]
Thanks for taking the time to consider and respond @joelostblom!

>  Could you give a rough indication of the amount of code that would be integrated into Altair, is it a few hundred lines or thousands?

I'd estimate around 500 lines of Python, plus tests. It would be a subset of the logic in https://github.com/hex-inc/vegafusion/tree/main/python/vegafusion/vegafusion.

**Comment by joelostblom** [comment]
And a follow up question, if we go this route, should the VegaFusion widget also be part of the extra_requires dependencies that are installed with `altair[full]` or is it meaningful to keep it as a separate package like the data server is kept separately now.

**Reply by jonmmease** [reply]
In the near term, it probably makes sense to keep `vegafusion-jupyter` as a separate package, and I could see including it in `altair[full]`.

**Comment by binste** [comment]
I've been using VegaFusion for a while now, mainly the mimetype renderer, and i also find it a very very "cool piece of technology"! :) It would be great to have it tighter integrated into Altair so that many more users can reap the benefits of this. +1 on everything said so far (with the caveat regarding enabling it by default, see below) and thank you @jonmmease for offering to do the integration! The example integration sounds good to me.

I'm wondering how we can enable the mimetype renderer by default if VegaFusion is installed without breaking too much. I think it's definitely worth trying and it's something to be looked at during the implementation. But just as an example, let's say an application creates an Altair chart, converts it to json with `chart.to_json()`, and then passes that Vega-Lite spec to a frontend so it can be rendered (think of e.g. Streamlit). With VegaFusion enabled, the json produced by `chart.to_json()` is currently still a VL spec but data is replaced with a vegafusion dataset URI which is no longer a self-contained VL spec that can be rendered. If VegaFusion would do the full transformation it would need to return a Vega spec which I'd find an unexpected behavior. Maybe a global "off" switch is enough with a future warning for a few releases if `chart.to_json/to_dict()` is used without explicitly specifying the `vegafusion` argument. I think it's definitely worth trying, would be great if it can be done!

In any case, very much looking forward to this! Happy to support with the implementation and review.

**Reply by jonmmease** [reply]
Thanks for the kind words @binste!

Yeah, using VegaFusion by default may not be a good idea (especially to start with). I think having it be an option that you turn on (either globally or at the renderer level) would be a great start.  Describing how to enable it in the `MaxRowsError` message would still make it pretty discoverable for those who would benefit!

**Comment by jonmmease** [comment]
@mattijn, do you have thoughts on the topic?

**Reply by mattijn** [reply]
First, with standard usage you might never see this max 5000 row error.

But I think we should add a table below the conda/pip install altair section, https://altair-viz.github.io/getting_started/installation.html, where we add the names of the soft dependencies with a description of their capabilities and why one want to install these (eg vl-convert-python, pyarrow, altair_viewer, vegafusion, any missing?).

We can assist power users though:
1. Place it a bit higher on https://altair-viz.github.io/user_guide/large_datasets.html, its completely at the bottom currently. 
2. I've you've installed it, we could allow for a opt-in or a opt-out behavior if you reach the max 5000 row error. 
    - For the opt-in, the error-message can be adapted that vegafusion is available. 
    - For the opt-out, vegafusion is auto-loaded, but there is a warning message stating that vegafusion was loaded and how to disable it (maybe warning message can disappear if you specifically import and enable it beforehand).

**Reply by mattijn** [reply]
Thanks for starting the discussion by the way @jonmmease!

**Reply by binste** [reply]
I like the idea of the table and bumping VegaFusion higher up in the docs! However, I think VegaFusion is also helpful if you don't get to the maxrowerror because it prunes unused columns. I have this regularly where I analyze a dataset with >20 columns, even if it just has 2000 rows, and the Jupyter notebook gets noticeably slower after creating a few charts which usually only reference ~3 columns. However, doing this with `import vegafusion as vf; vf.enable()` is very straightforward already.

As long as we can't enable it by default, which I currently believe to be the case, the main upside I see is that once vegafusion is installed, some functionalities could be provided without the user having to think about where it's coming from and in which documentation to look it up, e.g. the idea of having `chart.transformed_data()`.

The discussion feels similar to #2789. How should we deal with functionality which requires potentially too complex dependencies to be included in Altair? Make the separate packages optional dependencies and integrate them with Altair, merge them completely into Altair, or keep them separate and only reference in the documentation that they exist. The same question might come up at a later stage for https://github.com/binste/altair_basemap (probably soon to be renamed to `altair_tiles` and moved to `altair-viz` org). Looking at `pandas`, they have chosen the way of integrating many packages as [optional dependencies](https://pandas.pydata.org/docs/getting_started/install.html#optional-dependencies), the same goes for [matplotlib](https://matplotlib.org/stable/devel/dependencies.html#optional-dependencies). As a user, I personally like this approach as I have one documentation to read and one package to import and get prompted when I should install additional dependencies/add them to my requirements.

I think I don't have a good sense of the downsides of this. @mattijn where would you see potential issues?

**Reply by joelostblom** [reply]
> VegaFusion is also helpful if you don't get to the maxrowerror because it prunes unused columns.... the main upside I see is that once vegafusion is installed, some functionalities could be provided without the user having to think about where it's coming from and in which documentation to look it up

I think these are notable benefits as well and would love to have them more discoverable by users of altair.

> Looking at pandas, they have chosen the way of integrating many packages as [optional dependencies](https://pandas.pydata.org/docs/getting_started/install.html#optional-dependencies), the same goes for [matplotlib](https://matplotlib.org/stable/devel/dependencies.html#optional-dependencies). As a user, I personally like this approach as I have one documentation to read and one package to import and get prompted when I should install additional dependencies/add them to my requirements

I like this approach too as it seems to make it really easy for the user to choose if they want to take the "batteries included" route and install everything in one command or if they want to install only the core packages and then optional packages as needed. (A smaller point is how to name these, we could chose the command for installing everything to be the default `install altair` (together `install altair[core]` for no optional dependencies) or we could use `install altair[full]` for optional dependencies and `install altair` for the core functionality (it probably makes sense to go with whatever is the most common in other libraries for consistency).)

**Reply by mattijn** [reply]
I use Altair in the role of a data scientist, once I got the spec as prototype, a backend developer prepares tailor made (often aggregated) data by means of an API and there is a front-end developer who make sure the chart is reactive and embedded within an application and connect the params/signals to (styled) controls. In that setting I haven't had the need to use vegafusion extensively.

But I'm surely in favor of exploring further integration and I agree it should be optional. I like the suggested ideas how to integrate it more. Personally I favor the term `accelerator='vegafusion'` over `vegafusion=True`. 

How I have used (and liked) vegafusion was as a preprocessor. In this way I can connect and visualize large datafiles, collect the binned data using the `.transformed_data()`, build new dataframes and then use this transformed data as input for a chart without vegafusion enabled. That worked very well (but had to give up yesterday as I run into issues where the errors are currently a bit hard to trace.)

I am also interested how vegafusion can be further extended to give insight in chart details like one can do using the Vega editor (data-viewer, param-viewer). We discussed this recently in another issue how we can surface this information on the python side that is currently solely available within the Vega-editor.

I remember that the compilation into Vega instead of Vega-lite got me by surprise when opening the specification in Vega-editor and requires probably a bit more documentation. 

But all in all no negative feelings, quite the opposite!

**Reply by jonmmease** [reply]
Thanks for weighing in @mattijn, and thanks @joelostblom and @binste for the feedback as well. It sounds like we have good agreement on exploring the integration more.

As a next step, what would you all think about starting with the `chart.transformed_data()` method? This would involve:
 1. Porting over the [`evaluation.py`](https://github.com/hex-inc/vegafusion/blob/main/python/vegafusion/vegafusion/evaluation.py) functionality
 2. Porting over some testing logic.
 3. Adding VegaFusion as an optional dependency with a good error message when `transformed_data` is called but VegaFusion is not installed.
 4. Adding documentation.

Of this list, I could use the most help with the documentation plan, so if anyone has thoughts on where we would want to integrate transformed_data into the existing documentation structure, that would be great!

**Reply by mattijn** [reply]
Thanks @jonmmease, that sounds great! Let me also add you as maintainer to the altair-viz in case that makes things easier for the integration activities.

Two more cross-referencing issues:
1. In relation to surfacing the Vega datasets and signals: https://github.com/altair-viz/altair/issues/3040#issuecomment-1535507530
2. In relation to potential usage of Vegafusion for computing 2D density/(iso)contours https://github.com/altair-viz/altair/issues/891#issuecomment-1460147225. (not yet possible in vega-lite), (not necessarily in relation to `xarray`)

**Reply by joelostblom** [reply]
+1 to that plan! In terms of a place in the docs, I think it would be useful to add a section to the data [transformation landing page](https://altair-viz.github.io/user_guide/transform/index.html) on how to view transformed data. That section could start with a brief example of doing it with only Altair by [creating a table via mark_text](https://altair-viz.github.io/gallery/scatter_linked_table.html), and then show that it is much easier to have VegaFusion installed since you then can access `transformed_data` and essentially include the section you already have in the VegaFusion docs plus some more advanced example of what can be done now that we can programatically access the data. What do you think about that?

Another spot could be the ["Data transforms" section](https://altair-viz.github.io/user_guide/data_transformers.html), and I think maybe we should add something there about the new vegafusion transform options regardless.

To make the vegafusion advantages more discoverable, I think we should add sentences in 1-2 places in the docs that advertise these new sections (both for the transformed data and the fact that vf makes it easier to work with large datasets, but "by default" and via the widget if needed)

**Reply by binste** [reply]
+1 as well, having `.transformed_data` would be very useful and should not break anything for users so it's a great start. Looking forward to it and happy to review! And great to have you @jonmmease onboard as a maintainer, welcome :)

**Reply by jonmmease** [reply]
> Let me also add you as maintainer to the altair-viz in case that makes things easier for the integration activities.

Thanks @mattijn, I appreciate the trust.

Thanks for the documentation suggestions @joelostblom 

I'll keep you all posted as I start making progress on this!

**Reply by jonmmease** [reply]
Initial `transformed_data` PR: https://github.com/altair-viz/altair/pull/3081

**Reply by jonmmease** [reply]
merged, with `Chart._transformed_data()` as a private method for now. I'll start working on documentation updates over the next few days.

**Comment by jonmmease** [comment]
I've started playing some with how to integrate VegaFusion into the Altair rendering pipeline, and I think I have a pretty clean solution to propose.

First, add a new data transformer named "vegafusion". This data transformer will extract the dataframe-like data sources into a private `WeakValueDictionary` and replace them with special URLs that VegaFusion understands (basically `"table://{uuid}"`).

Other parts of Altair will check whether the active data transformer is named "vegafusion", and if so they will:
 1. compile the vega-lite spec to vega
 2. retrieve the extracted data sources form the private dictionary above
 3. Invoke VegaFusion's `pre_transform_spec` method with the Vega spec and the inline datasets to produce a new Vega spec with data transformations applied

(In practice this will all be wrapped up in a utility method)

So far, I've thought of three places to insert this logic:
 1. When rendering a chart (using any built-in renderer)
 2. When saving a chart (to any supported format)
 3. When converting a chart to a dictionary (`to_dict`) or JSON (`to_json`) if the `format` argument is set to `"vega"`. (If the format is `"vega-lite"`, then leave as-is).

A nice consequence of this approach is that a user will only need to run:
```python
alt.data_transformers.enable("vegafusion")
```
And then Altair will automatically use VegaFusion everywhere it makes sense.

Let me know what you think! If this general direction makes sense I'll work on a PR and we can discuss more there. Thanks!

**Reply by mattijn** [reply]
Sounds good! I like the approach of using an `altair.data_transformer()` here and I understand that you will have to introduce some logic at a few places to accommodate for this data-transformer.

**Comment by dangotbanned** [comment]
I think this discussion leads into this upstream discussion, and eventual release of [VegaFusion 2.0](https://vegafusion.io/posts/2024/2024-11-13_Release_2.0.0.html)
- https://github.com/vega/vegafusion/discussions/433
